### PR TITLE
Make Freddy JRuby compatible

### PR DESCRIPTION
**Pull request by madis**
This development will solve issues in using freddy with JRuby
where because of Threads and no GIL. It does it currently by using
another AMQP implementation - March Hare. It is specific to JRuby.

Because of the differences between Bunny & March Hare, some adapter
code is needed. E.g. AdaptiveQueue and Producer.

After these changes 2 gems would have to be built from this source:
  - freddy (ruby implementation)
  - freddy-jruby (jruby implementation)

By first making the Freddy compatible with JRuby will hopefully help to
discover concurrency issues and fix them so that Freddy can be used with
other Ruby implmemetations that have true concurrency e.g. Rubinius.

### CODE REVIEW

File: lib/freddy/adaptive_queue.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+class Freddy
+  class AdaptiveQueue
+      def initialize(queue)
```
Comment (urmastalimaa) [inline lib/freddy/adaptive_queue.rb]: Too deep indentation

File: lib/freddy/adaptive_queue.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        @queue = queue
+      end
+
+      def subscribe(&block)
```
Comment (urmastalimaa) [inline lib/freddy/adaptive_queue.rb]: What about creating a `FreddyQueue` implementation with abstract `subscribe` and then creating two subclasses, one for bunny and one for march_hare? The old *replace `if` with polymorphism* refacto.

**Comment by madis** [comment]
I'm in process of rebasing this branch to master as it was created off from old commit. Force push will come in next 30 min

File: lib/freddy/message_handler.rb [hunk 3]
```diff
@@ -4,17 +4,17 @@ class MessageHandler
 
     def initialize(adapter, delivery)
       @adapter = adapter
-      @properties = delivery.properties
-      @destination = @properties[:destination]
-      @correlation_id = @properties[:correlation_id]
+      @metadata = delivery.metadata
+      #@destination = @metadata.destination
```
Comment (urmastalimaa) [inline lib/freddy/message_handler.rb]: commented out code

File: lib/freddy/message_handlers.rb [hunk 4]
```diff
@@ -13,9 +13,8 @@ def initialize(producer, logger)
       def handle_message(payload, msg_handler, &block)
         block.call payload, msg_handler
       rescue Exception => e
-        destination = msg_handler.destination
-        @logger.error "Exception occured while processing message from #{destination}: #{Freddy.format_exception(e)}"
-        Freddy.notify_exception(e, destination: destination)
+        @logger.error "Exception occured while processing message from TODO: #{Freddy.format_exception(e)}"
```
Comment (urmastalimaa) [inline lib/freddy/message_handlers.rb]: TODO

**Comment by urmastalimaa** [comment]
Oops, I did not notice the WIP. I will continue with this when it's ready.

File: lib/freddy/responder_handler.rb [hunk 5]
```diff
@@ -19,11 +19,12 @@ def destroy_destination
     end
 
     def join
-      @channel.work_pool.join
+      # @channel.work_pool.join
```
Comment (indrekj) [inline lib/freddy/responder_handler.rb]: lets get rid of this

File: lib/freddy/responder_handler.rb [hunk 6]
```diff
@@ -19,11 +19,12 @@ def destroy_destination
...
     end
 
     def shutdown
-      @channel.work_pool.shutdown
+      # @channel.work_pool.shutdown
+      # @channel.close
```
Comment (indrekj) [inline lib/freddy/responder_handler.rb]: and this

Comment (indrekj) [inline lib/freddy/responder_handler.rb]: please check the whole class. maybe we don't need any of this

File: lib/freddy/message_handlers.rb [hunk 7]
```diff
@@ -44,7 +43,7 @@ def handle_message(payload, msg_handler, &block)
         end
       rescue Exception => e
         @logger.error "Exception occured while handling the request with correlation_id #{@correlation_id}: #{Freddy.format_exception(e)}"
-        Freddy.notify_exception(e, destination: msg_handler.destination, correlation_id: @correlation_id)
```
Comment (urmastalimaa) [inline lib/freddy/message_handlers.rb]: Why was this removed? It's quite useful information. If march_hare does not have that key then it can probably just be given as a constructor argument for MessageHandler.

File: lib/freddy/adaptive_queue.rb (hunk 2 above)
Comment (urmastalimaa) [inline lib/freddy/adaptive_queue.rb]: What did you think of this idea?

**Comment by indrekj** [comment]
@madis Can you look over the last few comments. Would be nice to merge it.

**Comment by madis** [comment]
Updated

**Comment by urmastalimaa** [comment]
+1
